core/src/com/unciv/ui/utils/KeyPressDispatcher.kt

```diff
@@ -61,14 +82,97 @@ class KeyPressDispatcher: HashMap<KeyCharAndCode, (() -> Unit)>() {
 
+    // access by KeyCharAndCode
+    operator fun set(key: KeyCharAndCode, action: () -> Unit) {
+        super.put(key, action)
+        checkInstall()
+    }
+    override fun remove(key: KeyCharAndCode): (() -> Unit)? {
+        val result = super.remove(key)
+        checkInstall()
+        return result
+    }
+
+    override fun toString(): String {
+        return (if (name==null) "" else "$name.") +
+            "KeyPressDispatcher(" + keys.joinToString(limit = 6){ it.toString() } + ")"
+    }
+
+    /** Removes all of the mappings, including a checkpoint if set. */
     override fun clear() {
         checkpoint = setOf()
         super.clear()
+        checkInstall()
     }
+
+    /** Set a checkpoint: The current set of keys will not be removed on a subsequent [revertToCheckPoint] */
     fun setCheckpoint() {
         checkpoint = keys.toSet()
     }
+    /** Revert to a checkpoint: Remove all mappings except those that existed on a previous [setCheckpoint] call.
+     * If no checkpoint has been set, this is equivalent to [clear] */
     fun revertToCheckPoint() {
         keys.minus(checkpoint).forEach { remove(it) }
+        checkInstall()
+    }
+
+    /** install our [EventListener] on a stage with optional inhibitor
+     * @param   stage The [Stage] to add the listener to
+     * @param   hasOpenPopups An optional lambda - when it returns true all keys are ignored
+     */
+    fun install(stage: Stage, name: String? = null, hasOpenPopups: (() -> Boolean)? = null) {
+        this.name = name
+        if (installStage != null) uninstall()
+        listener =
+            object : InputListener() {
+                override fun keyTyped(event: InputEvent?, character: Char): Boolean {
+                    val key = KeyCharAndCode(event, character)
+
+                    if (!contains(key) || (hasOpenPopups?.invoke() == true))
+                        return super.keyTyped(event, character)
+
+                    //try-catch mainly for debugging. Breakpoints in the vicinity can make the event fire twice in rapid succession, second time the context can be invalid
+                    try {
+                        this@KeyPressDispatcher[key]?.invoke()
+                    } catch (ex: Exception) {}
+                    return true
+                }
+            }
+        installStage = stage
+        checkInstall()
     }
+
+    /** uninstall our [EventListener] from the stage it was installed on. */
+    fun uninstall() {
+        checkInstall(forceRemove = true)
+        listener = null
+        installStage = null
+    }
+
+    /** Implements lazy hooking of the listener into the stage.
+     *
+     *  The listener will be added to the stage's listeners only when and as soon as [this] contains mappings.
+     *  When all mappings are removed or cleared the listener is removed from the stage.
+     */
+    private fun checkInstall(forceRemove: Boolean = false) {
+        if (listener != null && installStage != null) {
```

Comment:
Please reverse if to exit quickly and remove indentation :)

Reply:
Another personal preference - return before the end of a function still 'tastes' a bit like a goto - and G0t0 was not the best NPC in KOTOR2... 
Uh - I notice I left some `println`'s active - shouldn't have been. They have been really useful, but should be under a compiler conditional - and I haven't learned those for this ecosystem. `#ifdef` ? :wink: